src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/CalculationTransactionalService.kt

```diff
@@ -39,18 +42,117 @@ class CalculationTransactionalService(
   private val prisonService: PrisonService,
   private val domainEventPublisher: DomainEventPublisher,
   private val prisonApiDataMapper: PrisonApiDataMapper,
-  private val calculationService: CalculationService
+  private val calculationService: CalculationService,
+  private val bookingService: BookingService,
+  private val validationService: ValidationService,
 ) {
 
   fun getCurrentAuthentication(): AuthAwareAuthenticationToken =
     SecurityContextHolder.getContext().authentication as AuthAwareAuthenticationToken?
       ?: throw IllegalStateException("User is not authenticated")
 
+  /*
+   * There are 4 stages of full validation:
+   * 1. validate user input data and the raw sentence and offence data from NOMIS
+   * 2. Validate the transformation of the raw Booking (NOMIS offence data transformed into a Booking object)
+   * 3. Run the calculation and catch any errors thrown by the calculation algorithm
+   * 4. Validate the post calculation Booking (The Booking is transformed during the calculation). e.g. Consecutive sentences (aggregates)
+   */
+  fun fullValidation(prisonerId: String, calculationUserInputs: CalculationUserInputs): List<ValidationMessage> {
+    val sourceData = prisonService.getPrisonApiSourceData(prisonerId)
+    var messages = validationService.validateSourceData(sourceData) // Validation stage 1 of 4
+    if (messages.isNotEmpty()) return messages
+
+    val booking = bookingService.getBooking(sourceData, calculationUserInputs)
+    messages = validationService.validateBookingBeforeCalculation(booking) // Validation stage 2 of 4
+    if (messages.isNotEmpty()) return messages
+    try {
+      val bookingAfterCalculation = calculationService.calculate(booking) // Validation stage 3 of 4
+      validationService.validateBookingAfterCalculation(bookingAfterCalculation) // Validation stage 4 of 4
+    } catch (validationException: CrdCalculationValidationException) {
+      return listOf(ValidationMessage(validationException.validation))
+    }
+    return messages
+  }
+
+  //  The activeDataOnly flag is only used by a test endpoint (1000 calcs test, which is used to test historic data)
+  @Transactional
+  fun calculate(prisonerId: String, calculationUserInputs: CalculationUserInputs, activeDataOnly: Boolean = true): CalculatedReleaseDates {
+    val sourceData = prisonService.getPrisonApiSourceData(prisonerId, activeDataOnly)
+    val booking = bookingService.getBooking(sourceData, calculationUserInputs)
+    try {
+      return calculate(booking, PRELIMINARY, sourceData, calculationUserInputs)
+    } catch (error: Exception) {
+      recordError(booking, sourceData, calculationUserInputs, error)
+      throw error
```

Comment:
I'm not sure this will work. Throwing the error will roll back the transaction.

Reply:
may be misunderstanding this is what it used to do before? to discuss

Reply:
Sure, just realised there is no integration test for this. We should add one.

Reply:
add int test and refactor (as next PR)

Reply:
did a quick teston this by fircing an exception - it works. However TODO writer a proper test to test this (next PR)